libr/anal/block.c

```diff
@@ -0,0 +1,582 @@
+	if (block->addr == addr) {
+		r_anal_block_set_size (block, size);
+		return true;
+	}
+	if (r_anal_get_block_at (block->anal, addr)) {
+		// Two blocks at the same addr is illegle you know...
+		return false;
+	}
+
+	// Update the block's function's cached ranges
+	RAnalFunction *fcn;
+	RListIter *iter;
+	r_list_foreach (block->fcns, iter, fcn) {
+		if (fcn->meta._min != UT64_MAX) {
+			if (fcn->meta._max == block->addr + block->size) {
+				fcn->meta._max = addr + size;
+			}
+			if (fcn->meta._min == block->addr) {
+				fcn->meta._min = addr;
+			}
+		}
+	}
+
+	r_rbtree_aug_delete (&block->anal->bb_tree, &block->addr, __bb_addr_cmp, NULL, NULL, NULL, __max_end);
+	block->addr = addr;
+	block->size = size;
+	r_rbtree_aug_insert (&block->anal->bb_tree, &block->addr, &block->_rb, __bb_addr_cmp, NULL, __max_end);
+	return true;
+}
+
+R_API RAnalBlock *r_anal_block_split(RAnalBlock *bbi, ut64 addr) {
+	RAnal *anal = bbi->anal;
+	r_return_val_if_fail (bbi && addr >= bbi->addr && addr < bbi->addr + bbi->size && addr != UT64_MAX, 0);
+	if (addr == bbi->addr) {
+		return bbi;
+	}
+
+	if (r_anal_get_block_at (bbi->anal, addr)) {
+		// can't have two bbs at the same addr
+		return NULL;
+	}
+
+	// create the second block
+	RAnalBlock *bb = block_new (anal, addr, bbi->addr + bbi->size - addr);
+	if (!bb) {
+		return NULL;
+	}
+	bb->jump = bbi->jump;
+	bb->fail = bbi->fail;
+	bb->conditional = bbi->conditional;
+	bb->parent_stackptr = bbi->stackptr;
+
+	// resize the first block
+	r_anal_block_set_size (bbi, addr - bbi->addr);
+	bbi->jump = addr;
+	bbi->fail = UT64_MAX;
+	bbi->conditional = false;
+
+	// insert the second block into the tree
+	r_rbtree_aug_insert (&anal->bb_tree, &bb->addr, &bb->_rb, __bb_addr_cmp, NULL, __max_end);
+
+	// insert the second block into all functions of the first
+	RListIter *iter;
+	RAnalFunction *fcn;
+	r_list_foreach (bbi->fcns, iter, fcn) {
+			r_anal_function_add_block (fcn, bb);
+	}
+
+	// recalculate offset of instructions in both bb and bbi
+	int i;
+	i = 0;
+	while (i < bbi->ninstr && r_anal_bb_offset_inst (bbi, i) < bbi->size) {
+		i++;
+	}
+	int new_bbi_instr = i;
+	if (bb->addr - bbi->addr == r_anal_bb_offset_inst (bbi, i)) {
+		bb->ninstr = 0;
+		while (i < bbi->ninstr) {
+			ut16 off_op = r_anal_bb_offset_inst (bbi, i);
+			if (off_op >= bbi->size + bb->size) {
+				break;
+			}
+			r_anal_bb_set_offset (bb, bb->ninstr, off_op - bbi->size);
+			bb->ninstr++;
+			i++;
+		}
+	}
+	bbi->ninstr = new_bbi_instr;
+	r_anal_block_unref (bb);
+	return bb;
+}
+
+R_API bool r_anal_block_merge(RAnalBlock *a, RAnalBlock *b) {
+	r_anal_block_check_invariants (a->anal);
+	if (a->addr + a->size != b->addr) {
+		return false;
+	}
+
+	// check if function lists are identical
+	RAnalFunction *fcn;
+	RListIter *iter;
+	r_list_foreach (a->fcns, iter, fcn) {
+		if (!r_list_contains (b->fcns, fcn)) {
+			return false;
+		}
+	}
+	r_list_foreach (b->fcns, iter, fcn) {
```

Comment:
can't you just do:
```
if (r_list_length (a->fcns) != r_list_length (b->fcns)) {
	return false;
}

r_list_foreach (a->fcns, iter, fcn) {
	if (!r_list_contains (b->fcns, fcn)) {
		return false;
	}
}
```
and avoid the double loop?

Reply:
Yes, since the fcns list must not contain duplicates, that works. Thans!